### ISSUE TITLE
Large websocket responses get truncated and cannot be parsed

### ISSUE THREAD

**Issue by krassowski** [krassowski/jupyterlab-lsp#450]
## Description

Sometimes a very long response from the language server gets truncated and the relevant action is not performed. This appears to be somewhat random, because subsequent attempts can sometimes succeed.

```R
Uncaught SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at i.readMessage (index.js:formatted:1)
    at index.js:formatted:1
    at WebSocket.e.onmessage (index.js:formatted:1)
readMessage @ index.js:formatted:1
(anonymous) @ index.js:formatted:1
e.onmessage @ index.js:formatted:1
```

The message truncates at less than 7k characters

Similar issue was reported in a different project using PHP backend here: https://github.com/beyondcode/dusk-dashboard/issues/20.

## Reproduce

1. Open an R notebook
2. Type `base::` (potentially press <kbd>Tab</kbd>, but should auto-invoke)
4. See lack of completion suggestions and error in console

## Expected behavior

1. We may want to try to recover truncated JSON messages by parsing the viable portion (e.g. for completion it would make sense to show something rather than nothing, especially for extra long namespaces such as `base` in R)
2. The error should not happen in the first place

## Context

- Operating System and version: Ubuntu 20.10
- Browser and version: Chrome
- JupyterLab version: 2.2.x

master on 2.1.2

**Comment by bollwyvl** [comment]
Hm... We could also consider compression... Maybe have a configurable
threshold, as it's not free, and we wouldn't want it all the time. I'll
take a look.

**Comment by krassowski** [comment]
I wonder if this comment on the need of yielding when using `write_message` is relevant: https://github.com/tornadoweb/tornado/issues/1758#issuecomment-229977715?

**Comment by bollwyvl** [comment]
Hah, yeah, that would also do it.... Certainly worth a look.

**Comment by krassowski** [comment]
This is not a websocket issue. I tracked it down to `LspStdIoReader` which uses Python buffers to send messages received on stdio in a stream and for me:

```python
>>> import io
>>> io.DEFAULT_BUFFER_SIZE
8192
```

**Comment by krassowski** [comment]
And it also affects almost all of completion for javascript/typescript servers as they append huge amounts of metadata.

**Comment by krassowski** [comment]
For R language server:
- content-length header: 149007
- actual message intercepted by `raw = self.stream.read(content_length)`: 69606, truncated at `b'e":"function","packa'`

For javascript:
- content-length header: 180336
- actual message 65536 or 65648

The answer is here:

> If the argument is positive, and the underlying raw stream is not interactive, multiple raw reads may be issued to satisfy the byte count (unless EOF is reached first). But for interactive raw streams, at most one raw read will be issued, and a short result does not imply that EOF is imminent.

**Comment by dpriedel** [comment]
I believe I'm using the fixed code from above.  I have jupyter-lsp 1.1.1 and jupyterlab-lsp 3.2.0

I'm seeing this truncated output with the Jedi language server, pretty much all the time.

By way of an experiment, I changed line 122 in the _read_content method of the stdio.py
from 'break' to 'continue'.

The problem went away.

I wonder if the problem is that this async consumer routine is reading output from an async producer routine which just is not ready with its next chunk of data -- jedi language server not being the fastest tool in the shed.

So, maybe changing the 'break' to 'continue' and max_parts to some smaller number is a better way to go ??

I'm going to keep this change for a while and see how it goes.

Thanks

**Comment by krassowski** [comment]
This is very useful feedback, thank you! Please do keep us updated.

We could have `max_empty_parts_in_a_row=100` (or a different number) and only break if we cannot get a non-empty follow-up response for as many times. Would you like to create a PR?

**Comment by dpriedel** [comment]
I'll try using my hack for a day or so and if I don't run into any problems, I'll create a PR

**Comment by krassowski** [comment]
Fixed in #494.
